### PR TITLE
Use `std::from_chars` instead of `std::stoi`

### PR DESCRIPTION
**Pull request by glebm**
Adds a `ParseInt` function which uses `std::from_chars`.

From `std::from_chars` documentation:
> Unlike other parsing functions in C++ and C libraries,
> std::from_chars is locale-independent, non-allocating, and non-throwing.

### CODE REVIEW

File: Source/debug.cpp [hunk 1]
```diff
@@ -505,7 +532,8 @@ std::string DebugCmdQuest(const string_view parameter)
 
 std::string DebugCmdLevelUp(const string_view parameter)
 {
-	int levels = std::max(1, atoi(parameter.data()));
+	const ParseIntResult<int> parsedArg = ParseInt<int>(parameter, /*min=*/1);
+	const int levels = parsedArg.ok() ? parsedArg.value : 1;
```
Comment (ephphatha) [inline Source/debug.cpp:536]: `ParsedIntResult<T>::value_or(T value_if_error)` would be a great helper for cases like this, matches https://en.cppreference.com/w/cpp/utility/expected/value_or if c++23 support becomes a possibility :D

Comment (glebm) [reply Source/debug.cpp]: Done! We might as well add `tl::expected` once we've switched to C++17

File: Source/control.cpp [hunk 2]
```diff
@@ -411,10 +412,15 @@ std::string TextCmdArenaPot(const string_view parameter)
 		StrAppend(ret, _("Arenas are only supported in multiplayer."));
 		return ret;
 	}
+	int numPots = 1;
+	const ParseIntResult<int> parsedParam = ParseInt<int>(parameter, /*min=*/1);
+	if (parsedParam.ok()) {
+		numPots = parsedParam.value;
+	}
```
Comment (ephphatha) [inline Source/control.cpp:419]:
```suggestion
	int numPots = ParseInt<int>(parameter, /*min=*/1).value_or(1);
```

File: Source/debug.cpp [hunk 3]
```diff
@@ -191,8 +192,11 @@ std::string DebugCmdTakeGoldCheat(const string_view parameter)
...
 {
 	Player &myPlayer = *MyPlayer;
-	auto level = atoi(parameter.data());
-	if (level < 0 || level > (gbIsHellfire ? 24 : 16))
+	const ParseIntResult<int> parsedParam = ParseInt<int>(parameter, /*min=*/0);
+	if (!parsedParam.ok())
+		return "Specify the level number";
+	const int level = parsedParam.value;
+	if (level > (gbIsHellfire ? 24 : 16))
 		return StrCat("Level ", level, " is not known. Do you want to write a mod?");
```
Comment (ephphatha) [inline Source/debug.cpp:200]: This one is getting into the weeds a bit but given status allows discriminating between failure cases you can keep the old behaviour of returning one message for out-of-range values while also putting a distinct message for missing/not a number values.
```suggestion
	const ParseIntResult<int> parsedParam = ParseInt<int>(parameter, /*min=*/0, /*max=*/(gbIsHellfire ? 24 : 16));
	if (!parsedParam.ok()) {
		switch (parsedParam.status) {
		case ParseIntStatus::ParseError:
			return "Specify the level number";
		case ParseIntStatus::OutOfRange:
			return StrCat("Level ", level, " is not known. Do you want to write a mod?");
		}
	}
	const int level = parsedParam.value;
```

There's quite a few places where the value is later checked if it's in bounds which could use a similar approach. I'd leave this for a later refactor (I'm not really happy about using a switch for this...), it'd be a nice activity if `ParseIntResult` is made more like a `std::expected` specialisation.

**Review by ephphatha — COMMENTED** [review]
One other potential use of `value_or()`, the second comment is just a thought about error handling that doesn't warrant changing now. Consider this approval with a chance to accept the first suggestion :D

File: Source/debug.cpp (hunk 3 above)
Comment (glebm) [reply Source/debug.cpp:200]: This is a non-exhaustive switch, so it will result a warning.
Let's keep further improvements for a separate PR!

**Comment by glebm** [comment]
Rebased

**Comment by glebm** [comment]
@AJenbo Rebased on top of development and removed pre-C++17 fallback

**Pull request merged**